library/common/http/client.cc

```diff
@@ -599,18 +599,23 @@ const char* BaseCluster = "base";
 const char* H2Cluster = "base_h2";
 const char* H3Cluster = "base_h3";
 const char* ClearTextCluster = "base_clear";
+const char* ClearTextH2Cluster = "base_h2_clear";
 
 } // namespace
 
 void Client::setDestinationCluster(Http::RequestHeaderMap& headers) {
   // Determine upstream cluster:
   // - Use TLS with ALPN by default.
   // - Use http/2 or ALPN if requested explicitly via x-envoy-mobile-upstream-protocol.
-  // - Force http/1.1 if request scheme is http (cleartext).
+  // - Force http/1.1 if request scheme is http (cleartext) unless the protocol is set to h2c.
   const char* cluster{};
   auto protocol_header = headers.get(ProtocolHeader);
   if (headers.getSchemeValue() == Headers::get().SchemeValues.Http) {
-    cluster = ClearTextCluster;
+    if (!protocol_header.empty() && protocol_header[0]->value().getStringView() == "h2c") {
```

Comment:
Just want to call out that while any value for the header can be theoretically interpreted here, the iOS and Android public interfaces actually restrict setting this header, instead exposing a RequestBuilder option to set the HTTP protocol.

If the intent is to call this from iOS/Android you'll likely want to add an option to the respective RequestBuilders.